Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,746 @@
+    function ElementAt(index : (Int, Int), array : [|'T|]) {
+        return array[index];
+    }
+}
+example <'T> [||'T||] is IndexableBy<(Int, Int, Int), 'T> {
+    function ElementAt(index : (Int, Int, Int), array : [|'T|]) {
+        return array[index];
+    }
+}
+// This allows us to use the concept to write out more
+// general functions acting on arrays of different
+// dimensionality.
+function Subarray<'Array, 'Index, 'Element where 'Array is IndexableBy<'Index, 'Element>>(
+    array : 'Array,
+    indices : ['Index]
+) : 'Element[] {
+    mutable subarray = EmptyArray<'Element>();
+    for (index in indices) {
+        // We can use ElementAt to handle
+        // arrays of different dimensionality in
+        // a uniform fashion.
+        set subarray += ElementAt(index, array);
+    }
+    return subarray;
+}
+
+// Similarly, if we use type-parameterized UDTs and
+// discriminated unions, we can define a concept that
+// specifies when a value of a given type can possibly
+// be converted to another type.
+concept 'TOutput is MaybeConvertableFrom<'TInput> when {
+    function MaybeAs(input : 'TInput) : Maybe<'TOutput>;
+}
+// With that concept in place, we can say that some jagged
+// arrays can be converted to rectangular arrays, obviating
+// the need for JaggedAsRectangular2, JaggedAsRectangular3,
+// and so forth.
+example <'T> [|'T|] is MaybeConvertableFrom<[['T]]> when {
+    function MaybeAs(input : [['T]]]) : [|'T|] {
+        body intrinsic;
+    }
+}
+example <'T> [||'T||]] is MaybeConvertableFrom<[[['T]]]> when {
+    function MaybeAs(input : [[['T]]]]) : [||'T||] {
+        body intrinsic;
+    }
+}
+```
+
+### Alternatives
+
+#### Syntactic Sugar for Jagged Arrays
+
+As an alternative, one could consider not adding any new types to Q#, but rather extending `w/` to act on values of type `'T[][]`, `'T[][][]` using tuple indices:
+
+```qsharp
+// Construct the elementary matrix 𝑒₀₁ = |0⟩⟨1| as the jagged array
+// [[0, 1, 0], [0, 0, 0], [0, 0, 0]].
+let e01 = ConstantArray(3, ConstantArray(3, 0.0)) w/ (0, 1) <- 1.0;
+```
+
+#### Alternative Syntax for Multidimensional Literals
+
+As an alternative, one could consider keeping the new types and index expressions suggested by this proposal, but modifying the definition of multidimensional literals.
+
+For example, alternative delimiters could be used to separate rows, planes, and so forth:
+
+```qsharp
+let data3 = [
+    0, 1;  // Separate rows with ;
+    2, 3;; // Separate planes with ;;
+
+    4, 5;
+    6, 7
+];
+// data3: [||Int||]
+```
+
+Other delimiters than `;` could be considered to avoid overloading statement separators.
+For example, it may be considered less confusing to reuse the `|` character from `?|` expressions or the `\` character from string escapes:
+
+```qsharp
+let x = [
+    0.0, 1.0 |
+    1.0, 0.0
+];
+
+let z = [
+    1.0, 0.0 \\ // LaTeX-style notation
+    0.0, -1.0
+];
+```
+
+These alternatives can possibly be combined with `[| |]`:
```

Comment:
It is not clear to me how the snippet below illustrate how `[| |]` can be combined with the alternative of using `;` to terminate a row.

Reply:
Agreed, I think that made more sense with earlier versions of t he proposal... I'll go on and drop this comment, given that `[| |]` doesn't really combine well with `[ ...; ... ]`.